connaisseur/sigstore_validator.py

```diff
@@ -0,0 +1,76 @@
+import json
+import logging
+import re
+import subprocess  # nosec
+
+from connaisseur.crypto import decode_and_verify_ecdsa_key
+from connaisseur.exceptions import (
+    CosignError,
+    CosignTimeout,
+    NotFoundException,
+    ValidationError,
+)
+
+
+def get_cosign_validated_digests(image: str, pubkey: str):
+    """
+    Gets and processes cosign validation output for a given `image` and `pubkey`
+    and either returns a list of valid digests or raises a suitable exception
+    in case no valid signature is found or cosign fails.
+    """
+    failed, stdout, stderr = invoke_cosign(image, pubkey)
+    logging.info(
+        "COSIGN output for image: %s; RETURNCODE: %s; STDOUT: %s; STDERR: %s",
+        image,
+        failed,
+        stdout,
+        stderr,
+    )
+    digests = []
+    if not failed:
+        for sig in stdout.splitlines():
+            try:
+                sig_data = json.loads(sig)
+                digests.append(sig_data["Critical"]["Image"]["Docker-manifest-digest"])
```

Comment:
This means digests will be a list with entries of the form `sha256:<hash>`. The digests we obtain from notary are handled as plain `<hash>` without `sha256`. So, if I am not mistaken the image name is going to  be changed to `<image>@sha256:sha256:<hash>` [here](https://github.com/sse-secure-systems/connaisseur/blob/master/connaisseur/image.py#L86) which I would assume to result in an invalid image name, which cannot be pulled by k8s.

Reply:
Yep, seems so. Connaisseur admits this image in my test setup but then the image cannot be pulled:

```
  Warning  InspectFailed  4s (x3 over 6s)  kubelet, minikube  Failed to apply default image tag "docker.io/securesystemsengineering/testimage@sha256:sha256:c5327b291d702719a26c6cf8cc93f72e7902df46547106a9930feda2c002a4a7": couldn't parse image reference "docker.io/securesystemsengineering/testimage@sha256:sha256:c5327b291d702719a26c6cf8cc93f72e7902df46547106a9930feda2c002a4a7": invalid reference format
  Warning  Failed         4s (x3 over 6s)  kubelet, minikube  Error: InvalidImageName
```

Reply:
oh wow...how could I miss this :disappointed: great catch!!! Will fix it

Reply:
besides me being feeling bad for missing this, we should start checking in the integration tests whether we actually derived a valid digest